File: deepola/tpch_queries/q5.py
from deepola.operations import *
from deepola.query.query import Query

sql_query = """
select
	nation.n_name,
	sum(lineitem.l_extendedprice * (1 - lineitem.l_discount)) as revenue
from
	customer,
	orders,
	lineitem,
	supplier,
	nation,
	region
where
	customer.c_custkey = orders.o_custkey
	and lineitem.l_orderkey = orders.o_orderkey
	and lineitem.l_suppkey = supplier.s_suppkey
	and customer.c_nationkey = supplier.s_nationkey
	and supplier.s_nationkey = nation.n_nationkey
	and nation.n_regionkey = r_regionkey
	and region.r_name = 'ASIA'
	and orders.o_orderdate > date '1994-03-15'
	and orders.o_orderdate < date '1995-03-15'
group by
	nation.n_name
order by
	revenue desc;
"""

q = Query()
q.add_operation(name='table_customer',operation=TABLE(args={'table': 'customer'}))
q.add_operation(name='table_orders',operation=TABLE(args={'table': 'orders'}))
q.add_operation(name='table_lineitem',operation=TABLE(args={'table': 'lineitem'}))
q.add_operation(name='table_supplier',operation=TABLE(args={'table': 'supplier'}))
q.add_operation(name='table_nation',operation=TABLE(args={'table': 'nation'}))
q.add_operation(name='table_region',operation=TABLE(args={'table': 'region'}))

q.add_operation(name='where_o_orderdate_1',operation=WHERE(args={'form':'DNF','predicates': [[{'left':'o_orderdate','op':'>','right':'1994-03-15'}]]}))
q.add_operation(name='where_o_orderdate_2',operation=WHERE(args={'form':'DNF','predicates': [[{'left':'o_orderdate','op':'<','right':'1995-03-15'}]]}))
q.add_operation(name='where_r_name',operation=WHERE(args={'form':'DNF','predicates': [[{'left':'r_name','op':'==','right':'ASIA'}]]}))

q.add_operation(name='join_nation_region',operation=INNERJOIN(args={'left_on':['n_regionkey'],'right_on':['r_regionkey']}))
q.add_operation(name='join_supplier_nation',operation=INNERJOIN(args={'left_on':['s_nationkey'],'right_on':['n_nationkey']}))
q.add_operation(name='join_customer_supplier',operation=INNERJOIN(args={'left_on':['c_nationkey'],'right_on':['s_nationkey']}))
q.add_operation(name='join_customer_order',operation=INNERJOIN(args={'left_on':['c_custkey'],'right_on':['o_custkey']}))
q.add_operation(name='join_lineitem_order',operation=INNERJOIN(args={'left_on':['l_orderkey'],'right_on':['o_orderkey']}))
q.add_operation(name='join_lineitem_supplier',operation=INNERJOIN(args={'left_on':['l_suppkey'],'right_on':['s_suppkey']}))

q.add_operation(node_type="DM",name='groupby_operation',operation=GROUPBYAGG(args={'groupby_key':['n_name'],'aggregates':[{'op':'sum','col':'l_extendedprice*(1-l_discount)','alias':'revenue'}]}))
q.add_operation(node_type="DM",name='orderby_operation',operation=ORDERBY(args=[{'column':'revenue','order':'desc'}]))

q.add_operation(output=True,name='select_operation',operation=SELECT(args={'columns':['n_name','revenue']}))

q.add_edge('table_orders','where_o_orderdate_1')
q.add_edge('where_o_orderdate_1','where_o_orderdate_2')
q.add_edge('table_region','where_r_name')

q.add_edge('table_nation','join_nation_region')
q.add_edge('where_r_name','join_nation_region')

q.add_edge('table_supplier','join_supplier_nation')
q.add_edge('join_nation_region','join_supplier_nation')

q.add_edge('table_customer','join_customer_supplier')
q.add_edge('join_supplier_nation','join_customer_supplier')

q.add_edge('join_customer_supplier','join_customer_order')
q.add_edge('where_o_orderdate_2','join_customer_order')

q.add_edge('table_lineitem','join_lineitem_order')
q.add_edge('join_customer_order','join_lineitem_order')

q.add_edge('join_lineitem_order','join_lineitem_supplier')
q.add_edge('join_supplier_nation','join_lineitem_supplier')

q.add_edge('join_lineitem_supplier','groupby_operation')
q.add_edge('groupby_operation','orderby_operation')
q.add_edge('orderby_operation','select_operation')
q.compile()
# q.display()

File: rust/runtime/examples/polars-tpch/q5.rs
use crate::utils::*;
use std::collections::HashMap;
use polars::prelude::*;
use polars_lazy::prelude::*;

// select
// 	n_name,
// 	sum(l_extendedprice * (1 - l_discount)) as revenue
// from
// 	customer,
// 	orders,
// 	lineitem,
// 	supplier,
// 	nation,
// 	region
// where
// 	c_custkey = o_custkey
// 	and l_orderkey = o_orderkey
// 	and l_suppkey = s_suppkey
// 	and c_nationkey = s_nationkey
// 	and s_nationkey = n_nationkey
// 	and n_regionkey = r_regionkey
// 	and r_name = 'ASIA'
// 	and o_orderdate >= date '1994-01-01'
// 	and o_orderdate < date '1994-01-01' + interval '1' year
// group by
// 	n_name
// order by
// 	revenue desc;

pub fn read_polars_df(tablename: &str, tableinput: &HashMap<String, TableInput>, table_columns: &HashMap<String,Vec<&str>>) -> polars_lazy::frame::LazyFrame {
    let input_files = tableinput.get(tablename).unwrap().input_files.clone();
    let schema_cols = tpch_schema(tablename).unwrap().columns.iter().map(|x| x.name.clone()).collect::<Vec<String>>();
    let mut df = polars::prelude::CsvReader::from_path(input_files[0].clone()).unwrap()
    .has_header(false)
    .with_delimiter(b'|')
    .finish().unwrap();
    df.set_column_names(&schema_cols).unwrap();
    // Project columns
    df.lazy().select(&table_columns.get(tablename).unwrap().iter().map(|x| col(x)).collect::<Vec<Expr>>())
}

pub fn query(tableinput: HashMap<String, TableInput>) {
    // Create a HashMap that stores table name and the columns in that query.
    let table_columns = HashMap::from([
        ("lineitem".into(), vec!["l_orderkey","l_suppkey","l_extendedprice","l_discount"]),
        ("orders".into(), vec!["o_orderkey","o_custkey","o_orderdate"]),
        ("customer".into(), vec!["c_custkey","c_nationkey"]),
        ("supplier".into(), vec!["s_suppkey","s_nationkey"]),
        ("nation".into(), vec!["n_nationkey","n_name","n_regionkey"]),
        ("region".into(), vec!["r_regionkey","r_name"]),
    ]);

    let region_df = read_polars_df("region", &tableinput, &table_columns).filter(col("r_name").eq(lit("ASIA")));
    let nation_df = read_polars_df("nation", &tableinput, &table_columns);
    let nation_region_df = region_df.inner_join(nation_df, col("r_regionkey"), col("n_regionkey"));

    let supplier_df = read_polars_df("supplier", &tableinput, &table_columns);
    let supplier_nation_df = nation_region_df.inner_join(supplier_df, col("n_nationkey"), col("s_nationkey"));

    let lineitem_df = read_polars_df("lineitem", &tableinput, &table_columns);
    let lineitem_supplier_df = lineitem_df.inner_join(supplier_nation_df, col("l_suppkey"), col("s_suppkey"));

    let orders_df = read_polars_df("orders", &tableinput, &table_columns).filter(col("o_orderdate").gt_eq(lit("1994-01-01"))).filter(col("o_orderdate").lt(lit("1995-01-01")));
    let lineitem_orders_df = lineitem_supplier_df.inner_join(orders_df, col("l_orderkey"), col("o_orderkey"));

    let customer_df = read_polars_df("customer", &tableinput, &table_columns);
    let orders_customer_df = lineitem_orders_df.inner_join(customer_df, col("o_custkey"), col("c_custkey"));

    let result_df = orders_customer_df
        .groupby([col("n_name")])
    //     // Perform Aggregates
        .agg([
    //         col("l_quantity").sum().alias("sum_qty"),
    //         col("l_extendedprice").sum().alias("sum_base_price"),
            (col("l_extendedprice") * (lit(1) - col("l_discount"))).sum().alias("revenue"),
    //         (col("l_extendedprice") * (lit(1) - col("l_discount")) * (lit(1) + col("l_tax"))).sum().alias("sum_charge"),
    //         col("l_quantity").mean().alias("avg_qty"),
    //         col("l_extendedprice").mean().alias("avg_extendedprice"),
    //         col("l_discount").mean().alias("avg_disc"),
    //         col("l_linestatus").count().alias("count_order")
        ])
    //     // ORDER BY
        .sort_by_exprs(vec![col("revenue")], vec![false]);

    println!("Result: {:?}", result_df.collect());
}